handshake_server.go

```diff
@@ -299,11 +304,48 @@ Curves:
 		c.sendAlert(alertInternalError)
 		return false, err
 	}
-	if hs.clientHello.scts && hs.hello != nil {
+
+	// Set the private key for this handshake to the certificate's secret key.
+	hs.privateKey = hs.cert.PrivateKey
+
+	if hs.clientHello.scts {
 		hs.hello.scts = hs.cert.SignedCertificateTimestamps
 	}
 
-	if priv, ok := hs.cert.PrivateKey.(crypto.Signer); ok {
+	// Set the private key to the DC private key if the client and server are
+	// willing to negotiate the delegated credential extension.
+	//
+	// Check to see if a DelegatedCredential is available and should be used.
+	// If one is available, the session is using TLS >= 1.2, and the client
+	// accepts the delegated credential extension, then set the handshake
+	// private key to the DC private key.
+	if c.config.GetDelegatedCredential != nil && hs.clientHello.delegatedCredential {
```

Comment:
I think there is an update needed for the documentation and TLS 1.3. Spec says
```
   The DelegatedCredential contains a signature from the public key in
   the end-entity certificate using a signature algorithm advertised by
   the client in the "signature_algorithms" extension.  Additionally,
   the credential's public key MUST be of a type that enables at least
   one of the supported signature algorithms.  A Delegated credential
   MUST NOT be negotiated by the server if its signature is not
   compatible with any of the supported signature algorithms or the
   credential's public key is not usable with the supported signature
   algorithms of the client, even if the client advertises support for
   delegated credentials.
```

So if delegated credential returns signature type which is different than server public key type or is not on a ``signature_algorithms`` list, then it is a bug.
But TLS 1.3 introduces new extension called ``signature_algotirhms_cert`` (see 4.2.3 of a spec). It is a list of algorithms that client is able to verify. I think this implementation should look at ``signature_algotirhms_list`` if it is available and ``signature_algorithms`` in the other case.

Reply:
I'm not sure I understand. A correct implementation of `GetDelegatedCredential` will return a DC that uses a signature algorithm listed in `ClientHelloInfo.SignatureSchemes`. The question, then, is how is this list created. Per draft >= 23, this must be same as the list in the `signature_algorithms` extension.

Here's where I don't follow. The `signature_algorithms_cert` extension specifies the list of algorithms the client supports _for the purpose of verifying the server's cert chain._ If the client specifies this extension, then the `signature_algorithms` list applies only to verifying the handshake. (If it unspecified, then `signature_algorithms` applies to both.) So I don't see how there could be a bug here.

Unless the semantics of `ClientHelloInfo.SignatureSchemes` is changing in #102. Is that the case?

Reply:
So, in TLS 1.3:
* ``signature_algorithms_cert``: list of algorithms client is able to use to verify certificates (or server in case of client verification)
* ``signature_algorithms``: list of algorithms client is able to use to verify handshake signature. More specifically this list is used for CertificateVerify message only ...
* ... except if - ``signature_algorithms_cert`` is not provided  - then ``signature_algorithms`` list is used

Reply:
Right. So the semantics of `ClientHelloInfo.SignatureSchemes` in #102 is wrong. Effectively, there is one set of signature schemes the server uses to pick its certificate. If the `signature_algorithms_cert` extension is provided, then that's the set; otherwise the set provided in `signature_algorithms` is the set.

Reply:
Nop, code sets SignatureSchemes to list from which server needs to choose ceritficates. If TLS version is 1.3 AND ``signature_algorithms_cert`` is provided ( which corresponds to ``hs.clientHello.supportedSignatureAlgorithmsCert`` variable), then this set will be used. Otherwise ``hs.clientHello.supportedSignatureAlgorithms``

Reply:
See comment thread in #102.